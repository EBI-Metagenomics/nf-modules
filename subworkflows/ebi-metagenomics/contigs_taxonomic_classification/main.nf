include { DIAMOND_BLASTP  } from '../../../modules/ebi-metagenomics/diamond/blastp/main'
include { CATPACK_CONTIGS } from '../../../modules/ebi-metagenomics/catpack/contigs/main'

workflow CONTIGS_TAXONOMIC_CLASSIFICATION {
    take:
    contigs     // [ val(meta), path(assembly_fasta) ]
    proteins    // [ val(meta), path(proteins_fasta) ]
    cat_db      // [ val(meta), path(catdb_folder)  ]
    taxonomy_db // [ val(meta), path(cattax_folder) ]

    main:

    ch_versions = Channel.empty()

    /*
    * Note:
    * The CAT tool does not use the provided cat_db, as the alignment is performed by the Diamond step.
    * However, this option is mandatory in the CAT source code.
    */

    DIAMOND_BLASTP(
        proteins,
        [[id: "cat-db"], file("${cat_db[1]}/*.dmnd", checkIfExists: true)],
        6, // blast - txt
        []
    )
    ch_versions = ch_versions.mix(DIAMOND_BLASTP.out.versions.first())

    catpack_input_ch = contigs
        .join( proteins )
        .join( DIAMOND_BLASTP.out.txt )
        .multiMap { meta, contigs, proteins, diamond_txt ->
            contigs: [meta, contigs]
            proteins: [meta, proteins]
            diamond_txt: [meta, diamond_txt]
        }

    CATPACK_CONTIGS(
        catpack_input_ch.contigs,
        cat_db,
        taxonomy_db,
        catpack_input_ch.proteins,
        catpack_input_ch.diamond_txt
    )
    ch_versions = ch_versions.mix(CATPACK_CONTIGS.out.versions.first())

    emit:
    diamond_blast_tsv         = DIAMOND_BLASTP.out.txt
    contig2classification_tsv = CATPACK_CONTIGS.out.contig2classification
    versions                  = ch_versions
}
