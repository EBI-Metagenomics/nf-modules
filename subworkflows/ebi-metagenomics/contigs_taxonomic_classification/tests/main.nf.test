nextflow_workflow {

    name "Test subworkflow CONTIGS_TAXONOMIC_CLASSIFICATION"
    script "../main.nf"
    workflow "CONTIGS_TAXONOMIC_CLASSIFICATION"

    tag "subworkflows"
    tag "subworkflows_ebimetagenomics"
    tag "subworkflows/contigs_taxonomic_classification"
    tag "diamond"
    tag "diamond/blastp"
    tag "catpack/contigs"

    test("Run the pipeline on the test dataset") {
        when {
            params {
                output_path    = "${outputDir}"
            }
            workflow {
                """
                input[0] = Channel.of( [ [id:'test'], file("${moduleDir}/tests/data/contigs.fasta", checkIfExists: true) ] )
                input[1] = Channel.of( [ [id:'test'], file("${moduleDir}/tests/data/proteins.fasta", checkIfExists: true) ] )
                input[2] = [ [id:'test'], file("${moduleDir}/tests/data/db", checkIfExists: true) ]
                input[3] = [ [id:'test'], file("${moduleDir}/tests/data/tax", checkIfExists: true) ]
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
            )
        }
    }
}
