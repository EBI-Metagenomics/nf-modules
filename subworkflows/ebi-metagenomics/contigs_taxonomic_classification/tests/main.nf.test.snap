{
    "diamond_blast_tsv": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.txt:md5,d3f900af84ca793924006fd2923c2468"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-24T15:20:14.890404"
    },
    "contig2classification_tsv": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.contig2classification.txt:md5,a65813c054149642538768a45b57be0c"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-24T15:20:14.916609"
    },
    "html_contains_DOCTYPE_html_PUBLIC": {
        "content": [
            false
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-24T15:20:14.95627"
    },
    "versions": {
        "content": [
            [
                "versions.yml:md5,2325113dc53e34117c2db2c1a83a3197",
                "versions.yml:md5,2f5912bb61bbe6225306fdc5d263520a",
                "versions.yml:md5,356147dcf8c188490476c508a4c4c450",
                "versions.yml:md5,9d2a7059394e210cf1a76782967052ae"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-27T22:24:38.855938"
    }
}